test/type-applications/Main.hs

```diff
@@ -0,0 +1,443 @@
+import Control.Monad (guard)
+import Covenant.DeBruijn (DeBruijn (S, Z), asInt)
+import Covenant.Index
+  ( Index,
+    count0,
+    count1,
+    ix0,
+    ix1,
+  )
+import Covenant.Test (Concrete (Concrete))
+import Covenant.Type
+  ( AbstractTy,
+    BuiltinNestedT (ListT),
+    CompT (CompT),
+    Renamed (Rigid, Wildcard),
+    TypeAppError
+      ( DoesNotUnify,
+        ExcessArgs,
+        InsufficientArgs,
+        LeakingUnifiable
+      ),
+    ValT
+      ( Abstraction,
+        BuiltinNested,
+        ThunkT
+      ),
+    checkApp,
+    comp0,
+    comp1,
+    comp2,
+    integerT,
+    listT,
+    renameCompT,
+    renameValT,
+    runRenameM,
+    tyvar,
+    pattern ReturnT,
+    pattern (:--:>),
+  )
+import Data.Coerce (coerce)
+import Data.Functor.Identity (Identity (Identity))
+import Data.Kind (Type)
+import Data.Vector qualified as Vector
+import Test.QuickCheck
+  ( Gen,
+    Property,
+    arbitrary,
+    counterexample,
+    discard,
+    elements,
+    forAllShrink,
+    getSize,
+    liftShrink,
+    oneof,
+    shrink,
+    suchThat,
+    vectorOf,
+    (===),
+  )
+import Test.Tasty (adjustOption, defaultMain, testGroup)
+import Test.Tasty.HUnit (assertEqual, assertFailure, testCase)
+import Test.Tasty.QuickCheck (QuickCheckTests, testProperty)
+
+main :: IO ()
+main =
+  defaultMain . adjustOption moreTests . testGroup "Type application" $
+    [ testCase "HeadList on polymorphic empty list" unitLeakingUnifiable,
+      -- TODO: Check leaking wildcard; can't think of a case now
+      testProperty "Too many arguments to HeadList" propTooManyArgs,
+      testCase "HeadList on no arguments" unitInsufficientArgs,
+      testGroup
+        "Substitution"
+        [ testProperty "id applied to concrete" propIdConcrete,
+          testProperty "two-arg const to same concretes" propConst2Same,
+          testProperty "two-arg const to different concretes" propConst2Different
+        ],
+      testGroup
+        "Unification"
+        [ testProperty "concrete expected, concrete actual" propUnifyConcrete,
+          testProperty "unifiable expected, concrete actual" propUnifyUnifiableConcrete,
+          testProperty "rigid expected, concrete actual" propUnifyRigidConcrete,
+          testProperty "wildcard expected, concrete actual" propUnifyWildcardConcrete,
+          testProperty "concrete expected, unifiable actual" propUnifyConcreteUnifiable,
+          testProperty "unifiable expected, unifiable actual" propUnifyUnifiable,
+          testProperty "rigid expected, unifiable actual" propUnifyRigidUnifiable,
+          testProperty "wildcard expected, unifiable actual" propUnifyWildcardUnifiable,
+          testProperty "concrete expected, rigid actual" propUnifyConcreteRigid,
+          testProperty "unifiable expected, rigid actual" propUnifyUnifiableRigid,
+          testProperty "rigid expected, rigid actual" propUnifyRigid,
+          testProperty "wildcard expected, rigid actual" propUnifyWildcardRigid
```

Comment:
I believe a test case for a pair of `Abstraction`s is not covered :thinking: or a combination of abstraction and concrete type, etc.

Reply:
By this, do you mean something like `forall a b . a -> b -> (a, b)` together with `forall a . [a]` and `Integer`?

Reply:
Looking back at this, I believe I've meant some case that triggers the `expectPairOf` part of `unify` but not only with concrete types (as the Concrete arbitrary generates) as part of that pair but with e.g. `Abstraction`.

I think such case is missing in the suite and when testing against pair, there's only concrete types as of right now.

Reply:
Ah, I see! Yeah, that's a good catch. Will add such a case.